adds 1.24 cluster dockershim removal ready blog (#32303)
kubernetes/website: content/en/blog/_posts/2022-03-31-ready-for-dockershim-removal.md

---
layout: blog
title: "Is Your Cluster Ready for v1.24?"
date: 2022-03-31
slug: ready-for-dockershim-removal
---

**Author:** Kat Cosgrove

> Way back in December of 2020, Kubernetes announced the [deprecation of Dockershim](/blog/2020/12/02/dont-panic-kubernetes-and-docker/). In Kubernetes, dockershim is a software shim that allows you to use the entire Docker engine as your container runtime within Kubernetes. In the upcoming v1.24 release, we are removing Dockershim - the delay between deprecation and removal in line with the project’s policy of supporting features for at least one year after deprecation. If you are a cluster operator, this guide includes the practical realities of what you need to know going into this release. Also, what you need to do to make sure your cluster doesn’t fall over!

Reviewer: ```suggestion
Way back in December of 2020, Kubernetes announced the [deprecation of Dockershim](/blog/2020/12/02/dont-panic-kubernetes-and-docker/). In Kubernetes, dockershim is a software shim that allows you to use the Docker engine as your container runtime within Kubernetes. In the upcoming v1.24 release, we are removing Dockershim - the delay between deprecation and removal in line with the [project’s policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/) of supporting features for at least one year after deprecation. If you are a cluster operator, this guide includes the practical realities of what you need to know going into this release and what you need to do to make sure your cluster doesn’t fall over!
```

> Way back in December of 2020, Kubernetes announced the [deprecation of Dockershim](/blog/2020/12/02/dont-panic-kubernetes-and-docker/). In Kubernetes, dockershim is a software shim that allows you to use the entire Docker engine as your container runtime within Kubernetes. In the upcoming v1.24 release, we are removing Dockershim - the delay between deprecation and removal in line with the project’s policy of supporting features for at least one year after deprecation. If you are a cluster operator, this guide includes the practical realities of what you need to know going into this release. Also, what you need to do to make sure your cluster doesn’t fall over!

Reviewer 2: I'm (still) a bit wary of “we are removing Dockershim” when it's not Pulumi doing it. _I_ know what's meant; is there any appreciable risk of someone getting the wrong idea here?

Author: It doesn't give that impression to me, but if it's at all a risk I don't mind just removing my company affiliation from the blog. I only put it there because other blog authors tend to, anyway. :)

> 

Reviewer 2: > If you are a cluster operator, this guide includes the practical realities of what you need to know going into this release. 

This _isn't_ the release notes - I'm genuinely worried that someone will think that Kubernetes v1.24 is imminently due and that this is the entire upgrade guide.

Author: Reasonable -- I will rephrase this to avoid any confusion

If you are using a managed Kubernetes service from a cloud provider, and you haven’t explicitly changed the container runtime, it’s possible that there’s nothing else for you to do. Amazon EKS, Azure AKS, and Google GKE all default to containerd now, though if you have any node customizations, you should make sure they do not need updating. To check the runtime of your nodes, follow [Find Out What Container Runtime is Used on a Node](​​/docs/tasks/administer-cluster/migrating-from-dockershim/find-out-runtime-you-use/).

Regardless of whether you are rolling your own cluster or using a managed Kubernetes service from a cloud provider, you may need to [migrate telemetry or security agents that rely on Docker Engine](/docs/tasks/administer-cluster/migrating-from-dockershim/migrating-telemetry-and-security-agents/). 

## I have a Docker dependency. What now?

> If your Kubernetes cluster does depend on Docker Engine and you intend to upgrade to Kubernetes v1.24 (which you should do for security), you will need to either change your container runtime from Docker Engine to something else, or use [cri-dockerd](https://github.com/Mirantis/cri-dockerd). Since [containerd](https://containerd.io/) is a graduated CNCF project and the runtime within Docker itself, it’s a safe bet. Fortunately, the Kubernetes project has already documented the process of [changing a node’s container runtime](/docs/tasks/administer-cluster/migrating-from-dockershim/change-runtime-containerd/), using containerd as the example. Instructions are similar for switching to one of the other supported runtimes.

Reviewer 2: `suggestion
If your Kubernetes cluster does depend on Docker Engine and you intend to upgrade to Kubernetes v1.24 (which you should eventually do for security and similar reasons), you will need to either change your container runtime from Docker Engine to something else, or use [cri-dockerd](https://github.com/Mirantis/cri-dockerd). Since [containerd](https://containerd.io/) is a graduated CNCF project and the runtime within Docker itself, it’s a safe bet as an alternative container runtime. Fortunately, the Kubernetes project has already documented the process of [changing a node’s container runtime](/docs/tasks/administer-cluster/migrating-from-dockershim/change-runtime-containerd/), using containerd as the example. Instructions are similar for switching to one of the other supported runtimes.
```

> If your Kubernetes cluster does depend on Docker Engine and you intend to upgrade to Kubernetes v1.24 (which you should eventually do for security and similar reasons), you will need to either change your container runtime from Docker Engine to something else, or use [cri-dockerd](https://github.com/Mirantis/cri-dockerd). Since [containerd](https://containerd.io/) is a graduated CNCF project and the runtime within Docker itself, it’s a safe bet as an alternative container runtime. Fortunately, the Kubernetes project has already documented the process of [changing a node’s container runtime](/docs/tasks/administer-cluster/migrating-from-dockershim/change-runtime-containerd/), using containerd as the example. Instructions are similar for switching to one of the other supported runtimes.

Reviewer 3: ```suggestion
If your Kubernetes cluster depends on Docker Engine and you intend to upgrade to Kubernetes v1.24 (which you should eventually do for security and similar reasons), you will need to change your container runtime from Docker Engine to something else or use [cri-dockerd](https://github.com/Mirantis/cri-dockerd). Since [containerd](https://containerd.io/) is a graduated CNCF project and the runtime within Docker itself, it’s a safe bet as an alternative container runtime. Fortunately, the Kubernetes project has already documented the process of [changing a node’s container runtime](/docs/tasks/administer-cluster/migrating-from-dockershim/change-runtime-containerd/), using containerd as an example. Instructions are similar for switching to one of the other supported runtimes.
```

## I want to upgrade Kubernetes, and I need to maintain compatibility with Docker as a runtime. What are my options?

> Fear not, you aren’t being left out in the cold and you don’t have to take the security risk of staying on an old version of Kubernetes. Mirantis and Docker have jointly released, and are maintaining, a replacement for dockershim. That replacement is called [cri-dockerd](https://github.com/Mirantis/cri-dockerd). If you do need to maintain compatibility with Docker as a runtime, install cri-dockerd following the instructions in the project’s documentation.

Reviewer 3: ```suggestion
Fear not, you aren’t being left out in the cold and you don’t have to take the security risk of staying on an old version of Kubernetes. Mirantis and Docker have jointly released, and are maintaining, a replacement for dockershim. That replacement is called [cri-dockerd](https://github.com/Mirantis/cri-dockerd). If you need to maintain Docker compatibility as a runtime, install cri-dockerd following the instructions in the project’s documentation.
```

Reviewer 2: > If you need to maintain Docker compatibility as a runtime

To me that implies that we're calling the reader a runtime! I prefer the original wording.

Reviewer: Haha agreed.

## Is that it?

> Yes. As long as you go into this release aware of the changes being made and the details of your own clusters, as well as communicate clearly with your development teams, it will be minimally dramatic. You may have some changes to make to your cluster, your application code, or some scripts, but all of these requirements are documented. Switching from using Docker Engine as your runtime to using containerd effectively means removing the middleman, since containerd is the runtime used by Docker itself. From a practical perspective, this removal is better both for you and for Kubernetes maintainers in the long-run. If you still have questions, please first check the [Dockershim Removal FAQ](/blog/2022/02/17/dockershim-faq/).

Reviewer 3: ```suggestion
Yes. As long as you go into this release aware of the changes being made and the details of your own clusters and communicate clearly with your development teams, it will be minimally dramatic. You may have some changes to make to your cluster, application code, or scripts, but all of these requirements are documented. Switching from using Docker Engine as your runtime to using containerd effectively means removing the middleman since containerd is the runtime used by Docker itself. From a practical perspective, this removal is better both for you and for Kubernetes maintainers in the long-run. If you still have questions, please first check the [Dockershim Removal FAQ](/blog/2022/02/17/dockershim-faq/).
```

Reviewer 2: For clarity, maybe this:
```suggestion
Yes. As long as you go into this release aware of the changes being made and the details of your own clusters, and you make sure to communicate clearly with your development teams, it will be minimally dramatic. You may have some changes to make to your cluster, application code, or scripts, but all of these requirements are documented. Switching from using Docker Engine as your runtime to using containerd effectively means removing the middleman since containerd is the runtime used by Docker itself. From a practical perspective, this removal is better both for you and for Kubernetes maintainers in the long-run.

If you still have questions, please first check the [Dockershim Removal FAQ](/blog/2022/02/17/dockershim-faq/).
```

Reviewer: Do you think it makes sense to specifically call out containerd here? I would like to see us be less opinionated about which supported runtime we advocate and instead direct people to the list of supported options: https://kubernetes.io/docs/setup/production-environment/container-runtimes/.
